impeller/renderer/backend/vulkan/resource_manager_vk_unittests.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <sys/types.h>
+#include <functional>
+#include <memory>
+#include <utility>
+#include "fml/synchronization/waitable_event.h"
+#include "gtest/gtest.h"
+#include "impeller/renderer/backend/vulkan/resource_manager_vk.h"
+
+namespace impeller {
+namespace testing {
+
+// While expected to be a singleton per context, the class does not enforce it.
+TEST(ResourceManagerVKTest, CreatesANewInstance) {
+  auto const a = ResourceManagerVK::Create();
+  auto const b = ResourceManagerVK::Create();
+  EXPECT_NE(a, b);
+}
+
+namespace {
+
+// Invokes the provided callback when the destructor is called.
+//
+// Can be moved, but not copied.
+class DeathRattle final {
+ public:
+  explicit DeathRattle(std::function<void()> callback)
+      : callback_(std::move(callback)) {}
+
+  DeathRattle(DeathRattle&&) = default;
+  DeathRattle& operator=(DeathRattle&&) = default;
+
+  ~DeathRattle() { callback_(); }
+
+ private:
+  std::function<void()> callback_;
+};
+
+}  // namespace
+
+TEST(ResourceManagerVKTest, ReclaimMovesAResourceAndDestroysIt) {
+  auto const manager = ResourceManagerVK::Create();
+
+  auto waiter = fml::AutoResetWaitableEvent();
+  auto dead = false;
+  auto rattle = DeathRattle([&waiter]() { waiter.Signal(); });
+
+  // Not killed immediately.
+  EXPECT_FALSE(waiter.IsSignaledForTest());
+
+  {
+    auto resource = UniqueResourceVKT<DeathRattle>(manager, std::move(rattle));
+
+    // Not killed on moving.
+    EXPECT_FALSE(waiter.IsSignaledForTest());
```

Comment:
This can fail too.  I'd just remove this assert.  Once the rattle is passed to the ResourceManagerVK all bets are off about the order of operations.  It is possible that the thread running the test could pause and the ResourceManagerVK thread could get all the execution time which would allow it to call ~DeathRattle() before this executes.  It's an unlikely race but not impossible.

In order to keep the assert you'd need to add another `AutoResetWaitableEvent` and another `UniqueResourceVKT` that signals it so you can essentially pause the ResourceManagerVK until the assertion completes.

Reply:
Thanks for explaining. That doesn't seem worth it here, so removed.